### ISSUE TITLE
UI components that link to Property can result in memory leak

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/sun#121]
This is relevant to sun and scenery-phet.  

Some UI components link to one or more Property instances that are provided as constructor arguments (e.g., valueProperty) or options (e.g. enabledProperty).  That linking is done in the constructor.  If the UI component has a lifetime that's shorter than the Property that it's linking to, you have a memory leak.  What's needed is an API to tell the UI component to 'clean up' (see for example reactants-products-and-leftovers.IntegerSpinner.unlink).

Here's an incomplete list of UI components that have this problem:

sun
- [x] BooleanRectangularToggleButton
- [x] BooleanRoundStickyToggleButton
- [x] BooleanRoundToggleButton
- [x] ExpandCollapseButton
- [x] RadioButtonGroup (if options.enabledProperty is provide)
- [x] RectangularStickyToggleButton
- [x] RectangularToggleButton
- [x] RoundButtonView
- [x] RoundStickyToggleButton
- [x] RoundToggleButton
- [x] ToggleNode

scenery-phet
- [x] FaucetNode
- [x] GaugeNode
- [x] KitControlNodeSides
- [x] KitControlNodeTop
- [x] KitSelectionNode
- [x] LeftRightSpinner
- [x] MeasuringTape
- [x] NumberPicker
- [x] PlayPauseButton
- [x] RewindButton
- [x] ScientificNotationNode
- [x] SoundToggleButton
- [x] StepButton
- [x] ThermometerNode
- [x] TimerToggleButton
- [x] UpDownSpinner
- [x] WavelengthSlider

**Comment by pixelzoom** [comment]
Some of this is being addressed in https://github.com/phetsims/sun/issues/170.

**Comment by pixelzoom** [comment]
Also need to call tandem.removeComponent in any components that are instrumented with data collection.

**Comment by pixelzoom** [comment]
At the very least, this should be triaged before "Summer 2015 redeploy", to see if we may have sims that are leaking memory because of this issue.

**Comment by pixelzoom** [comment]
Assigning to @ariel-phet to decide how/whether to address this.

**Comment by pixelzoom** [comment]
5/28/15 dev meeting: Not necessary for "Summer 2015 redeploy". If a sim is leaking memory, it's OK if it continues to leak memory for this release. We'll probably address this when we start instrumenting sims for data collection, since that will require addressing the dispose issue for tandem.removeInstance.

**Comment by ariel-phet** [comment]
I kind of want to unassign this issue from myself...but it seems like we might want a new universal label such as "address during instrumentation" or such (but less verbose). Since it seems with instrumentation work there will be many associated sub issues

**Comment by samreid** [comment]
In my opinion, this issue would be discovered by a together utility that tracks instance counts during instrumentation, I created a together issue for it.  In my opinion, this issue can be unassigned since it is not important enough to be solved for S2015R.

**Comment by pixelzoom** [comment]
At 11/12/15 dev meeting, we decided:

(1) Adding `dispose` to sun components is important, because sun is supposed to be generally useful outside of PhET.  So we should be proactive about fixing them.

(2) PhET-specific common components (e.g. scenery-phet) will be fixed as we hit them.

**Comment by pixelzoom** [comment]
4/7/16 dev meeting:
Assign to @andrewadare, consult with @pixelzoom. Chip away at this gradually.  Main repos are sun and scenery-phet, and there are some component-specific issues. Use Chrome dev tools to test for memory leaks.  If sim leaks are noticed, create an issue for the sim.

**Comment by pixelzoom** [comment]
@andrewadare see scenery-phet.FaucetNode for a good example of the patterns that are typically employed in `dispose`.   The 3 areas of house-keeping are typically:

(1) remove observers from things (Property, Emitter, Events) that were provided by the client
(2) dispose of any subcomponents that require disposal
(3) de-registration from tandem (data collection)

**Comment by pixelzoom** [comment]
Also note in FaucetNode that `dispose` calls `disposeFaucetNode`, which is `@private` and does all of the work. This is a general pattern, designed to avoid having to create a lot of `this` fields for things that are needed solely by dispose. The name of the private function is always of the form `dispose + constructor.name`, in this example `dispose + FaucetNode`.

**Comment by pixelzoom** [comment]
4/5/18 dev meeting:
• pattern in https://github.com/phetsims/sun/issues/121#issuecomment-209141994 is still preferred
• some of the UI components identified here may have been fixed already
• @pixelzoom will work with @andrea-phet on this

**Comment by jonathanolson** [comment]
I'm not sure what to respond to (why it's tagged for the developer meeting), but if sun button work is being done, https://github.com/phetsims/sun/issues/362 could be done presumably at the same time.

**Comment by pixelzoom** [comment]
This is tagged for developer meeting to decide how to proceed with adding/fixing `dispose` for sun buttons.

**Comment by jbphet** [comment]
This has been low priority for such a long time because having buttons come and go from a sim is a rare occurrence, so there isn't a lot of real memory leakage occurring, at least not yet.  @pixelzoom mentioned that he had at least one sim where buttons *do* come and go (was it Models of the Hydrogen Atom?).  Perhaps one way to prioritize the work would be to look at this sim and any others where buttons come and go and make sure that the buttons used in those sims are addressed first.

As one of the original authors of button code, I'd be glad to work on this if the powers that be decide that it is a top priority.

**Comment by Denz1994** [comment]
@andrea-phet Will continue work and retag for dev meeting when this becomes blocking. @jbphet currently has other higher priorities.

**Comment by pixelzoom** [comment]
To clarify the above comment...  @andrea-phet will continue working on scenery-phet components, likely skipping anything named *Button.

Self un-assigning. @andrea-phet let me know if there's anything you'd like me to review.

**Comment by andrea-phet** [comment]
@pixelzoom, FaucetNode, GaugeNode, and MeasuringTape are ready for review.

**Comment by andrealin** [comment]
I have checked the following scenery-phew components for memory leaks and implemented/fix dispose where necessary:

- [ ] FaucetNode
- [ ] GaugeNode
- [ ] LeftRightSpinner
- [ ] MeasuringTape
- [ ] NumberPicker
- [ ] ScientificNotationNode
- [ ] ThermometerNode
- [ ] WavelengthSlider

**Comment by pixelzoom** [comment]
@andrealin I completed review of the components listed in https://github.com/phetsims/sun/issues/121#issuecomment-399162547.  Two minor changes in the commits above.  Let me know if you have any questions.

**Comment by andrealin** [comment]
The commits above make sense. Removing the status:block-publication label since all changes have been reviewed.

**Comment by andrealin** [comment]
Dispose has been implemented or fixed in the scenery-phet components in (list) that are not buttons. Retagging for developer meeting to evaluate how to proceed.

**Comment by andrealin** [comment]
Re assigning to myself to make sure the current scenery-phet components with dispose are not leaking memory.

**Comment by zepumph** [comment]
I helped out @chrisklus today and we found that after `RadioButtonGroupMember`'s interaction state property is disposed as part of https://github.com/phetsims/sun/issues/383, there doesn't look to be any more memory leaking in `RadioButtonGroupMember`.

**Comment by pixelzoom** [comment]
@chrisklus @zepumph serious problems in many of these dispose functions.  E.g.  ButtonModel:

```js
    this.disposeButtonModel = function() {

      // This will unlink all listeners, causing potential issues if listeners try to unlink properties afterwards
      this.overProperty.dispose();
      this.downProperty.dispose();
      this.enabledProperty.dispose();
      this.listeners = [];
    };
```

Each function has its own context, so `this` needs to be `self` inside of these functions.

This problem is present in these types (and possibly others):

ExpandCollapseButton
ButtonModel
MomentaryButtonModel
PushButtonModel
RadioButtonGroupMember
RectangularButtonView
RectangularPushButton
RoundButtonView

**Comment by pixelzoom** [comment]
... and this makes me wonder.  Are you actually testing these changes?...

**Comment by pixelzoom** [comment]
Mea culpa... In these functions, `this` refers to the owner of the function. So either `this` or `self` will work.   But it's still confusing to use both `self` and `this`, e.g. in MomentaryButtonModel:

```js
    this.disposeMomentaryButtonModel = function() {
      self.downProperty.unlink( downListener );
      this.pressedEmitter.removeListener( setValueOn );
      this.releasedEmitter.removeListener( setValueOff );
    };
```

**Comment by chrisklus** [comment]
@pixelzoom yes of course I'm testing any changes, it's the only way to know if fixing a leak was successful. I agree that we should be consistent on using `self` or `this`, I'll look out for that in the future.

**Comment by chrisklus** [comment]
Update for 10/18/18 dev meeting:
@SaurabhTotey and I were able to check off 4 more that we were previously unable to figure out since they were caused by the memory leak in scenery, see https://github.com/phetsims/scenery/issues/879.

**Comment by chrisklus** [comment]
Update for 10/25/18 dev meeting:
@SaurabhTotey and I fixed three more this week. Next up: RectangularToggleButton

**Comment by chrisklus** [comment]
Update for 11/1/18 dev meeting:
@SaurabhTotey and I checked off four more this week. Next up: LeftRightSpinner

**Comment by chrisklus** [comment]
Update from 11/15/18 dev meeting: Pairing time for @SaurabhTotey and @chrisklus got moved to https://github.com/phetsims/states-of-matter/issues/217, but @SaurabhTotey is going to finish these remaining items on his own. Once the current list is finished, we'll bring back to dev meeting to evaluate if this issue is complete, or if more components need to be added and tested/fixed.

**Comment by chrisklus** [comment]
As decided in 2/7/19 dev meeting, we're going to close this issue and not take explicit action to test more components. Devs will keep an eye out for possibly leaking components when doing sim memory tests. Closing, cheers @SaurabhTotey!

**Comment by samreid** [comment]
During review of the PlayControlButton, we noticed this code:

```js
    isPlayingProperty.link( isPlayingListener );

    // @private
    this.disposePlayStopButton = () => {
      if ( isPlayingProperty.hasListener( isPlayingListener ) ) {
        isPlayingProperty.unlink( isPlayingListener );
      }
```

That is, the listener is always added, but conditionally removed. I noticed there are 3 commits in this issue with the commit message: 

> Check hasListener before unlinking, for phetsims/sun#121

Are all of those checks necessary?  Are some of them potentially hiding a problem?  We can open a new issue for investigation if desired but I thought I would start here for context.  @pixelzoom or @chrisklus can you please comment?

**Comment by pixelzoom** [comment]
> Are all of those checks necessary?

They were necessary at one time.  It used to be the case that `unlink` threw an assertion if the listener was not a listener.  So if you added a listener to a Property that you didn't own, you needed to check `hasListener` because the owner might have disposed that Property before you were disposed.

That apparently changed at some point, see TinyEmitter `removeListener`:

```js
    if ( assert && !this.isDisposed ) {
      assert( this.listeners.has( listener ), 'tried to removeListener on something that wasn\'t a listener' );
    }
```

> Are some of them potentially hiding a problem?

No, I don't think this is a problem - it's good defensive programming.  Imo the decision to make`unlink` be a no-op in some circumstances is much more likely to hide problems. But that's a decision I don't feel like revisiting.

**Comment by samreid** [comment]
Thanks for the explanation.  Also, I hadn't noticed that `isPlayingProperty` is passed in to PlayControlButton and hence is not owned.  Closing.
